### ISSUE TITLE
os.getenv("A_NUMBER") can return an int

### ISSUE THREAD

**Issue by dhalbert** [adafruit/circuitpython#9113]
If `settings.toml` contains:
```
A_NUMBER = 123
```
`os.getenv("A_NUMBER")`  will return the `int` `123`.

This is non-standard compared with CPython, where `os.getenv()` always returns a string or `None`.

This came up because I was thinking about adding TOML bool support (`true` and `false`) to what we can parse in `settings.toml`.

I'm not sure this is worth fixing, but I think it's possible to fix.

**Comment by jepler** [comment]
It's worth noting that the formula `x = int(getenv("VAR"))` will be portable between circuitpython & standard python.

**Comment by NateChurch** [comment]
I ran into a related problem because of my assumptions. 

in my settings.toml
```toml
MQTT_BROKER=mqtt.svc
```
I run 
```python
mqtt_broker = os.getenv("MQTT_BROKER")
```
and I get this error
```REPL
ValueError: invalid syntax for integer with base 10
```

If I alter my settings.toml
```toml
MQTT_BROKER="mqtt.svc"
```
then it runs fine. I had assumed it would always return a string but it doesn't. I also assumed it was trying to make a float because it has a period in it, but that isn't a thing either. It wants strings to have quotes.

**Comment by dhalbert** [comment]
@NateChurch TOML wants strings to have quotes: https://toml.io/en/

**Comment by dhalbert** [comment]
See
https://github.com/adafruit/circuitpython/pull/10422#issuecomment-2972938020:
from @Neradoc:

> If we are completely disallowing ints then it's a breaking change that we need to document clearly with the mention that we only accept strings in our subset of toml (and that they need double quotes).
> https://docs.circuitpython.org/en/latest/docs/environment.html#details-of-the-toml-language-subset
> ```
> • The supported data types are string and integer
> ...
> • Only integers supported by strtol. (no 0o, no 0b, no underscores 1_000, 011 is 9, not 11)
> ```
> Alternatively we could still allow reading ints but return them as strings (and have a better error) but that could be confusing.
> 
> Also, it looks like this preserves `common_hal_os_getenv_int`, but does that mean that the user code can't read some of the environment variables used by Circuitpython ?
> ```toml
> CIRCUITPY_WEB_API_PORT=8000
> CIRCUITPY_DISPLAY_WIDTH=640
> CIRCUITPY_DISPLAY_COLOR_DEPTH=1
> ```

and
https://github.com/adafruit/circuitpython/pull/10422#issuecomment-2972951411
from @jepler:

> I have no HW so I'm not able to test on HW, but under this PR the intent is that internal routines that get integers will continue to work as before.
> 
> It is only intended to change the os.getenv API so that it can no longer return integers (cpython compatibility) and will issue a better error message in the case that a non-quoted value is encountered with a key passed to os.getenv.
> 
> I'm not sure whether I'm in favor of this change myself, because as you say it ought to go through deprecation and does remove the ability to read integer-syntax values from settings.toml into python code without adding it back in some hidden way. It's just that I was still being copied on comments on https://github.com/adafruit/circuitpython/issues/9113 and noticed a recent one go by. I decided to check into the coding required to implement CPython compatibility and this is what I came up with. But I'm not actually itching to see it merged, to be honest.

**Comment by dhalbert** [comment]
The fundamental problem, as we know, is that TOML provides multiple data types, and `os.getenv()` is only strings.

I see some different possibilities:
1. Don't change the current non-CPython behavior.
2. Error out when doing `os.getenv()` on a non-string value. This is #10422 as of now.
3. Change `os.getenv()` to always return the value as a string. The user code can parse the string as appropriate. EDIT: I started to do this, just by changing https://github.com/adafruit/circuitpython/blob/325d49ce309de248e465345bae3d39fa6744f168/shared-module/os/getenv.c#L379-L383 to
```c
     return mp_obj_new_str_from_vstr(&buf); 
```
But this isn't quite right, because it picks up trailing spaces from things like `v = 7 # comment`.
4. Add something like `supervisor.get_setting(key)` which will return a typed value consistent with TOML typing.

3 and 4 could be done together.

**Comment by tannewt** [comment]
I agree that 3 and 4 would be good.

**Comment by BobCochran** [comment]
I'm an Adafruit customer and CircuitPython user. I'm happy to help test any solutions for this issue. Thank you so much!

**Comment by NateChurch** [comment]
> [@NateChurch](https://github.com/NateChurch) TOML wants strings to have quotes: https://toml.io/en/

Good point. I wasn't thinking of it as a _toml_, I was thinking of it as an _.env_ which of course is the problem you point out later in the thread. I realize that all of this was months ago, but wanted to chime in that I like 3 and 4 also. Those feel like the best logical solution to align with expectations and behaviors in big python. There could also be a **toml** library like in big python, but that feels like overkill for normal usage.

**Comment by dhalbert** [comment]
I am implementing 3 and 4 above:

### **os.getenv(key, default)**

- `os.getenv(key, default)` always returns a string, which is the value on the right side of the `key =` in `settings.toml`. Leading and trailing whitespace is removed. Surrounding double-quotes are removed. The value may be invalid in TOML, but no error is raised. If there is no `settings.toml` or the key is not found, the default value is returned.

### **supervisor.get_setting(key, default)**

`supervisor.get_setting(key, default)` returns a `str`, an `int`, or a `bool`. If the key is not found or `settings.toml` is not present, then the default value is returned. If the value is present but is malformed and cannot be parsed as one of those types, a `ValueError` is raised. (`KeyError` is not raised; this is more like `dict.get()`.

If you have an API suggestion other than the above, I'm interested in hearing it. This needs to be a small API so it can fit on small boards.
